src/audio/module_adapter/module/volume/volume.c

```diff
@@ -1090,7 +1090,8 @@ static int volume_process(struct processing_module *mod,
 			frames = avail_frames;
 		} else if (cd->ramp_type == SOF_VOLUME_LINEAR_ZC) {
 			/* with ZC ramping look for next ZC offset */
-			frames = cd->zc_get(input_buffers[0].data, cd->vol_ramp_frames, &prev_sum);
+			frames = cd->zc_get(input_buffers[0].cached_data, cd->vol_ramp_frames,
+					    &prev_sum);
```

Comment:
Should we always pass a cached ptr here and WB/INV if we pass between different cores ?

Reply:
@lgirdwood here it *is* always the cached pointer. And it is handled by the coherent API, so all the required cache synchronisation and locking is done there. Maybe what's a bit confusing is that even though it is a union and so the user can access the same field using either of the pointers, they must not be mixed. I suspect this might be confusing and error prone. Maybe we need an API like
```
void init_stream_buffers(input, n_input, output, n_output, cached)
{
	for (i = 0; i < n_input; i++)
		input[i].type = cached ? TYPE_CACHED : TYPE_UNCACHED;
	for (i = 0; i < n_output; i++)
		output[i].type = cached ? TYPE_CACHED : TYPE_UNCACHED;
}

int stream_buffer_input_cached(input, void __sparse_cache **data)
{
	if (input->type != TYPE_CACHED)
		return -EINVAL;
	*data = input->cached_data;
	return 0;
}
int stream_buffer_output_cached(output, void __sparse_cache **data) {...}
int stream_buffer_input_uncached(input, void **data) {...}
int stream_buffer_output_uncached(output, void **data) {...}
```

Reply:
Sorry, what I mean is we should always pass cache coherent cachable pointers to modules as part of process(). i.e the core module adaptor takes care of the WB/INV if the data passes core boundary.
This way the module has to take no action and treats the data as coherent (just as it would as a host application).

Reply:
@lgirdwood you mean not only for `.simple_copy` but for fully opaque third party modules too? Well, I don't know the module API well enough to make this call alone. Are modules guaranteed to execute their `.process()` fully synchronously on the same core? Is the API really supposed to manage data buffers for those modules or is it up to them to decide?

Reply:
Yes, modules will fully execute ```process()``` or ```copy()``` on a single core with this core affinity across the module instance lifetime. A module should have no knowledge of the cache model used by the underlying HW so the module adapter should take care of managing this in a platform specific manner. 
i.e. on Intel platforms
1) Module input buffers will be cache pointer types and module API will make sure contents are coherent prior to ```process()```
2) Module output buffers will be cache pointer types and module API will make sure contents are coherent after ```process()```

The module adapter will only WB/INV buffers if they will cross core boundaries or to/from DMA on Intel HW.

Reply:
@lgirdwood and `process()` is guaranteed to not spawn any asynchronous threads? To execute fully synchronously? Do we have any such module to test - I'd feel more comfortable doing such a change if I could test it. I understand adding this union with two types of pointers isn't very pretty, especially to then be removed again, but maybe still it would be better to do this in two steps for easier debugging? I see this rather risky with 3rd party modules.

Reply:
```process()``` will run in the same context as the caller i.e. in the same thread/core.

Reply:
@lgirdwood with a slight delay `.data` is now unconditionally cached

Reply:
@lyakh Does PR title require an update too?